### ISSUE TITLE
deno info --json does not include relevant redirection data

### ISSUE THREAD

**Issue by danopia** [denoland/deno#9351]
When a module imports a URL which returns a redirect, it's not possible to correlate the dependency from the info json. The __importing__ module has the __original__ URL listed as a dependency, while the __imported__ module is represented with only the __new__ URL.

Brief example:

`deno info --unstable --json https://gist.githubusercontent.com/danopia/d8b92fdbaa146133dac74a248e62d761/raw/bf5074703f24fee4c2f08577908115f2a6ffff6a/repro.ts`
```json
{
  "compiled": null,
  "depCount": 1,
  "fileType": "TypeScript",
  "files": {
    "https://deno.land/std@0.85.0/examples/cat.ts": {
      "deps": [],
      "size": 238
    },
    "https://gist.githubusercontent.com/danopia/d8b92fdbaa146133dac74a248e62d761/raw/bf5074703f24fee4c2f08577908115f2a6ffff6a/repro.ts": {
      "deps": [
        "https://deno.land/std/examples/cat.ts"
      ],
      "size": 48
    }
  },
  "local": "/home/dan/.cache/deno/deps/https/gist.githubusercontent.com/1825a41fb76917b8414a149f2a36b45ab33986ff6f1adbc14f514641720b278d",
  "map": null,
  "module": "https://gist.githubusercontent.com/danopia/d8b92fdbaa146133dac74a248e62d761/raw/bf5074703f24fee4c2f08577908115f2a6ffff6a/repro.ts",
  "totalSize": 286
}
```

If you search that JSON for `cat.ts` you'll see that there are exactly two mentions and they are for different URLs.

### possible fixes?

Given that multiple source URLs can redirect to the same destination URL, I don't think there's an easy fix that fits into the data model. Normal `deno info` only prints the source URL which isn't great because the actual loaded URL gets dropped on the ground. My better idea would be introducing a `redirections` string->string key to allow consumers to opt-in to accounting for redirects.

**Comment by kitsonk** [comment]
We should address this as part of #7927.

I am working on that for 1.8 will be breaking the JSON significantly because it doesn't really represent the graph.  I will try to post the proposed JSON schema soon and see if it addresses the concerns here, because I agree it isn't represented well at all.

**Comment by danopia** [comment]
> I will try to post the proposed JSON schema soon and see if it addresses the concerns here, because I agree it isn't represented well at all.

Sounds excellent, given that I just wrote [a tool consuming deno info --json](https://deno-visualizer.danopia.net/dependencies-of/https/github.com/ihack2712/texa-svelte-dev-starter/raw/master/server.ts?rankdir=LR) today and would love to actually have reliable redirection following & type-only deps. I look forward to it.

**Comment by kitsonk** [comment]
@danopia internally we keep a map of redirects, but I am not too sure it is really what we want to expose in the "graph".  The format is going to change significantly, and align much more to the way we represent it internally overall.

For example, I think it will look something like this:

```json
{
  "root": "https://deno.land/x/mod.ts",
  "modules": {
    "https://deno.land/x/mod.ts": {
      "dependencies": {
        "./a.js": {
          "code": "https://deno.land/x/a.js",
          "type": "https://deno.land/x/a.d.ts"
        }
      }
    },
    "https://deno.land/x/a.js": {},
    "https://deno.land/x/a.d.ts": {}
  }
}
```

Where each module is just a "key" on the modules, and you would walk it from the "root" module outwards.  If there was a re-direct, that would just be resolved in the dependency already, so that the `"code"` or `"type"` value would be the resolved module which would be listed in the graph.

**Comment by danopia** [comment]
Ok, that's an interesting example. So I imagine the file would look like this for my example gist:

```json
{
  "root": "https://gist.githubusercontent.com/danopia/.../repro.ts",
  "modules": {
    "https://gist.githubusercontent.com/danopia/.../repro.ts": {
      "dependencies": {
        "https://deno.land/std/examples/cat.ts": {
          "code": "https://deno.land/std@0.85.0/examples/cat.ts"
        }
      }
    },
    "https://deno.land/std@0.85.0/examples/cat.ts": {}
  }
}
```

This is an extrapolation, hopefully I conveyed correctly :) but given the understanding that the deps objects are keyed off the given import string, I'd be able to "opt into" knowing about redirections by checking if they are two different URLs, and thus indicate that an unversioned import is in use, for example. So that sounds reasonable for my usecase I think.

**Comment by kitsonk** [comment]
Yes, if you specified full URLs in the import, but if they were relative and redirected, you would only know about the resolved final URL, but I would argue that is all anyone should be concerned with anyways, because that is what we expect a module graph to do, resolve all specifiers to the files that get loaded.

**Comment by kitsonk** [comment]
I believe this is addressed in #9630 in the sense that only the finally resolved URL is provided.

**Comment by danopia** [comment]
> Brief example:
> 
> `deno info --unstable --json https://gist.githubusercontent.com/danopia/d8b92fdbaa146133dac74a248e62d761/raw/bf5074703f24fee4c2f08577908115f2a6ffff6a/repro.ts`
> 
> ```json
> {
>   "compiled": null,
>   "depCount": 1,
>   "fileType": "TypeScript",
>   "files": {
>     "https://deno.land/std@0.85.0/examples/cat.ts": {
>       "deps": [],
>       "size": 238
>     },
>     "https://gist.githubusercontent.com/danopia/d8b92fdbaa146133dac74a248e62d761/raw/bf5074703f24fee4c2f08577908115f2a6ffff6a/repro.ts": {
>       "deps": [
>         "https://deno.land/std/examples/cat.ts"
>       ],
>       "size": 48
>     }
>   },
>   "local": "/home/dan/.cache/deno/deps/https/gist.githubusercontent.com/1825a41fb76917b8414a149f2a36b45ab33986ff6f1adbc14f514641720b278d",
>   "map": null,
>   "module": "https://gist.githubusercontent.com/danopia/d8b92fdbaa146133dac74a248e62d761/raw/bf5074703f24fee4c2f08577908115f2a6ffff6a/repro.ts",
>   "totalSize": 286
> }
> ```


This now looks like this:

```json
{
  "root": "https://gist.githubusercontent.com/danopia/d8b92fdbaa146133dac74a248e62d761/raw/bf5074703f24fee4c2f08577908115f2a6ffff6a/repro.ts",
  "modules": [
    {
      "specifier": "https://deno.land/std@0.88.0/examples/cat.ts",
      "dependencies": [],
      "size": 238,
      "mediaType": "TypeScript",
      "local": "/home/dan/.cache/deno/deps/https/deno.land/763ed57a3356c55ee9018dcd426a777c067fd7e9f46cc4c24c5681241d1a2224",
      "checksum": "8fe4051f5dc6f0b01f1844605a5a6fe8e339cf2debc79069cf03fb59eef1eed4"
    },
    {
      "specifier": "https://gist.githubusercontent.com/danopia/d8b92fdbaa146133dac74a248e62d761/raw/bf5074703f24fee4c2f08577908115f2a6ffff6a/repro.ts",
      "dependencies": [
        {
          "specifier": "https://deno.land/std/examples/cat.ts",
          "isDynamic": false,
          "code": "https://deno.land/std@0.88.0/examples/cat.ts"
        }
      ],
      "size": 48,
      "mediaType": "TypeScript",
      "local": "/home/dan/.cache/deno/deps/https/gist.githubusercontent.com/1825a41fb76917b8414a149f2a36b45ab33986ff6f1adbc14f514641720b278d",
      "checksum": "39489aa25e3b5db53b4a3beca038b05b7441b18a27d2207a4fff24427431ec43"
    }
  ],
  "size": 286
}
```

LGTM
